### PR TITLE
modify kryo serialization. fix #2881

### LINKED ISSUE #2881: Kryo serialization fails（服务既是生产者又是消费者出现问题）
https://github.com/apache/incubator-dubbo/issues/2179#issue-347287850

传输的数据较大，想使用kryo序列化，来提高性能。请问，在最新的2.6.5版本下，还是会出现
image

请问大神们，这个问题要如何解决呢，谢谢！！！！

### PR DESCRIPTION
**Pull request by cvictory**
## What is the purpose of the change

fix #2881

## Brief changelog

先提交讨论下方案可行性。之后再来完善单元测试和部分优化点，不要进行merge。（Submit the feasibility of the program under discussion first. Then improve the unit test and partial optimization points.）

original  (原来的结构) :
image

now （修改之后）:
image

现在维护多个id和class列表。

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by carryxyh** [comment]
我认为这种修复方式是可以的。
我会发送一封邮件来获取更多的意见，这是一个比较大的改动，你认为呢？
另外，这个pr是否应该同样提交到master上？


I think this kind of modifications is ok.
I will send an email to get more comments. This is a big change. What do you think?
In addition, should this PR be submitted to the owner as well?

File: dubbo-serialization/dubbo-serialization-kryo/src/main/java/com/alibaba/dubbo/common/serialize/kryo/utils/MultiThreadLocalKryoFactory.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+package com.alibaba.dubbo.common.serialize.kryo.utils;
+
+import com.esotericsoftware.kryo.Kryo;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MultiThreadLocalKryoFactory extends ThreadLocalKryoFactory {
```
Comment (carryxyh) [inline dubbo-serialization/dubbo-serialization-kryo/src/main/java/com/alibaba/dubbo/common/serialize/kryo/utils/MultiThreadLocalKryoFactory.java]: Does the function of kryoUtils have a certain overlap with KryoFactory? Personally think that KryoUtils is a bit redundant. The current use of ThreadLocalKryoFactory in KryoUtils is also not flexible.

Comment (cvictory) [reply dubbo-serialization/dubbo-serialization-kryo/src/main/java/com/alibaba/dubbo/common/serialize/kryo/utils/MultiThreadLocalKryoFactory.java]: 是的，原来的调用链路比较长。我这个也是一个修改的初稿，等方案完全确定，再来做些修改和完善单元测试。（上周休假了）

Yes, the original call link is longer. I am also a revised draft, and the plan is completely determined, and then I will make some modifications and improve the unit test. (I took a vacation last week)

Comment (carryxyh) [reply dubbo-serialization/dubbo-serialization-kryo/src/main/java/com/alibaba/dubbo/common/serialize/kryo/utils/MultiThreadLocalKryoFactory.java]: Sorry, I forgot to send the email. I will send the email today. I think we can discuss it there.
:)

**Comment by khanimteyaz** [comment]
to me looks fine.

**Pull request closed without merging**